core/src/main/java/greencity/service/impl/TipsAndTricksServiceImpl.java

```diff
@@ -0,0 +1,167 @@
+
+@Service
+@RequiredArgsConstructor
+public class TipsAndTricksServiceImpl implements TipsAndTricksService {
+    @Value("${messaging.rabbit.email.topic}")
+    private String sendEmailTopic;
+
+    private final TipsAndTricksRepo tipsAndTricksRepo;
+
+    private final UserService userService;
+
+    private final ModelMapper modelMapper;
+
+    private final RabbitTemplate rabbitTemplate;
+
+    private final NewsSubscriberService newsSubscriberService;
+
+    private final TipsAndTricksTagsService tipsAndTricksTagsService;
+
+    private final FileService fileService;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public AddTipsAndTricksDtoResponse save(AddTipsAndTricksDtoRequest addTipsAndTricksDtoRequest, MultipartFile image,
+                                            String email) {
+        TipsAndTricks toSave = modelMapper.map(addTipsAndTricksDtoRequest, TipsAndTricks.class);
+        toSave.setAuthor(userService.findByEmail(email));
+        if (addTipsAndTricksDtoRequest.getImage() != null) {
+            image = convertToMultipartImage(addTipsAndTricksDtoRequest.getImage());
+        }
+        if (image != null) {
+            toSave.setImagePath(fileService.upload(image).toString());
+        }
+        toSave.setTipsAndTricksTags(addTipsAndTricksDtoRequest.getTipsAndTricksTags()
+            .stream()
+            .map(tipsAndTricksTagsService::findByName)
+            .collect(Collectors.toList())
+        );
+
+        try {
+            tipsAndTricksRepo.save(toSave);
+        } catch (DataIntegrityViolationException e) {
+            throw new NotSavedException(ErrorMessage.TIPS_AND_TRICKS_NOT_SAVED);
+        }
+
+        return modelMapper.map(toSave, AddTipsAndTricksDtoResponse.class);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PageableDto<TipsAndTricksDto> findAll(Pageable page) {
+        Page<TipsAndTricks> pages = tipsAndTricksRepo.findAllByOrderByCreationDateDesc(page);
+        List<TipsAndTricksDto> tipsAndTricksDtos = pages
+            .stream()
+            .map(tipsAndTricks -> modelMapper.map(tipsAndTricks, TipsAndTricksDto.class))
+            .collect(Collectors.toList());
+
+        return new PageableDto<>(
+            tipsAndTricksDtos,
+            pages.getTotalElements(),
+            pages.getPageable().getPageNumber()
+        );
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PageableDto<TipsAndTricksDto> find(Pageable page, List<String> tags) {
+        Page<TipsAndTricks> pages = tipsAndTricksRepo.find(page, tags);
+
+        List<TipsAndTricksDto> tipsAndTricksDtos = pages.stream()
+            .map(tipsAndTricks -> modelMapper.map(tipsAndTricks, TipsAndTricksDto.class))
+            .collect(Collectors.toList());
+
+        return new PageableDto<>(
+            tipsAndTricksDtos,
+            pages.getTotalElements(),
+            pages.getPageable().getPageNumber()
+        );
```

Comment:
Use private method for this from previous comment

Reply:
Done